### PR TITLE
Ports over LongRAGPack, Corrective RAG Pack, and Self-Discover Pack to Workflows

### PR DESCRIPTION
**Pull request by jonathanhliu21**
# Description

Changes the following LlamaPacks to use the new workflow feature, and adds examples under the workflow folder showcasing the usage of these workflows in the context of implementing the LlamaPacks.
- LongRAGPack
- Corrective RAG
- Self-Discover

## New Package?

Not a new package.

## Version Bump?

Did I bump the version in the `pyproject.toml` file of the package I am updating? (Except for the `llama-index-core` package)

- [x] Yes
- [ ] No

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Added new unit/integration tests
- [x] Added new notebook (that tests end-to-end)
- [x] I stared at the code and made sure it makes sense

## Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have added Google Colab support for the newly added notebooks.
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

**Comment by logan-markewich** [comment]
Awesome @jonathanhliu21 !

I'm curious if you had any overall feedback (good or bad) after trying out the workflows syntax? 👀

**Comment by jonathanhliu21** [comment]
> Awesome @jonathanhliu21 !
> 
> I'm curious if you had any overall feedback (good or bad) after trying out the workflows syntax? 👀

Overall, I think that structuring workflows is pretty intuitive. In terms of feedback, I think being able to specify which step to run when `run()` is called at the beginning could be helpful when there are multiple steps that can take in a `StartEvent`. Additionally, defining event classes was somewhat tedious in the self-discover case, as all of the variables in the event class was the same type. However, I wasn't completely sure if this is the right way to use workflows, so please let me know if there's a cleaner way to define events.

**Comment by masci** [comment]
> I think being able to specify which step to run when `run()` is called at the beginning could be helpful when there are multiple steps that can take in a `StartEvent`.

That's tricky because ideally the workflow shouldn't depend on the order of the processed events. In other words, if two steps wait for a `StartEvent`, which one goes first shouldn't matter. If it matters, the step that needs to go after shouldn't probably take a `StartEvent`.

For context, what you ask is definitely doable with the current asyncio implementation, but might not be the case the moment we offer a different workflows "runtime".

File: llama-index-packs/llama-index-packs-corrective-rag/llama_index/packs/corrective_rag/base.py [hunk 1]
```diff
@@ -44,89 +55,161 @@
...
+    async def retrieve(self, ctx: Context, ev: StartEvent) -> Optional[RetrieveEvent]:
         """Retrieve the relevant nodes for the query."""
-        retriever = self.index.as_retriever(**kwargs)
-        return retriever.retrieve(query_str)
-
-    def evaluate_relevancy(
-        self, retrieved_nodes: List[Document], query_str: str
-    ) -> List[str]:
+        query_str = ev.get("query_str")
+        retriever_kwargs = ev.get("retriever_kwargs")
```
Comment (logan-markewich) [inline llama-index-packs/llama-index-packs-corrective-rag/llama_index/packs/corrective_rag/base.py:111]: might be safer to do `retriever_kwargs = ev.get("retriever_kwargs", {})`

File: llama-index-packs/llama-index-packs-corrective-rag/llama_index/packs/corrective_rag/base.py [hunk 2]
```diff
@@ -44,89 +55,161 @@
...
+        ctx.data["tavily_tool"] = TavilyToolSpec(api_key=tavily_ai_apikey)
 
-    def get_modules(self) -> Dict[str, Any]:
-        """Get modules."""
-        return {"llm": self.llm, "index": self.index}
+        return StopEvent()
 
-    def retrieve_nodes(self, query_str: str, **kwargs: Any) -> List[NodeWithScore]:
+    @step(pass_context=True)
+    async def retrieve(self, ctx: Context, ev: StartEvent) -> Optional[RetrieveEvent]:
```
Comment (logan-markewich) [inline llama-index-packs/llama-index-packs-corrective-rag/llama_index/packs/corrective_rag/base.py:108]: I guess this event assumes the index is already constructed. Should we raise an error if not? Or make the index / llm / tavilly tool part of the workflow constructor?

File: llama-index-packs/llama-index-packs-corrective-rag/llama_index/packs/corrective_rag/base.py [hunk 3]
```diff
@@ -44,89 +55,161 @@
...
-        relevancy_results = self.evaluate_relevancy(retrieved_nodes, query_str)
-        # Extract texts from documents that are deemed relevant based on the evaluation.
-        relevant_text = self.extract_relevant_texts(retrieved_nodes, relevancy_results)
+class CorrectiveRAGPack(BaseLlamaPack):
+    def __init__(self, documents: List[Document], tavily_ai_apikey: str) -> None:
+        """Init params."""
+        self._wf = CorrectiveRAGWorkflow()
+
+        asyncio.run(
+            self._wf.run(documents=documents, tavily_ai_apikey=tavily_ai_apikey)
```
Comment (logan-markewich) [inline llama-index-packs/llama-index-packs-corrective-rag/llama_index/packs/corrective_rag/base.py:203]: ah I see, we have this in the constructor of the rag pack itself

File: llama-index-packs/llama-index-packs-corrective-rag/llama_index/packs/corrective_rag/base.py [hunk 4]
```diff
@@ -44,89 +55,161 @@
...
 
-        # Compile the final result. If there's additional search text from the transformed query,
-        # it's included; otherwise, only the relevant text from the initial retrieval is returned.
-        if search_text:
-            return self.get_result(relevant_text, search_text, query_str)
-        else:
-            return self.get_result(relevant_text, "", query_str)
+    def run(self, query_str: str, **kwargs: Any) -> Any:
+        """Run the pipeline."""
+        return asyncio.run(self._wf.run(query_str=query_str, retriever_kwargs=kwargs))
```
Comment (logan-markewich) [inline llama-index-packs/llama-index-packs-corrective-rag/llama_index/packs/corrective_rag/base.py:215]: small nit: slightly safer to use our util instead

```
from llama_index.core.async_utils import asyncio_run

...
return asyncio_run(...)
```

File: llama-index-packs/llama-index-packs-corrective-rag/pyproject.toml [hunk 5]
```diff
@@ -29,7 +29,7 @@ license = "MIT"
 maintainers = ["ravi-theja"]
 name = "llama-index-packs-corrective-rag"
 readme = "README.md"
-version = "0.1.1"
+version = "0.1.2"
```
Comment (logan-markewich) [inline llama-index-packs/llama-index-packs-corrective-rag/pyproject.toml:32]: we should bump the llama-index-core dep to be v0.10.61 as well

File: llama-index-packs/llama-index-packs-self-discover/llama_index/packs/self_discover/base.py [hunk 6]
```diff
@@ -49,96 +58,111 @@
...
+
+class SelfDiscoverWorkflow(Workflow):
+    """Self discover workflow."""
+
+    @step(pass_context=True)
+    async def get_modules(self, ctx: Context, ev: StartEvent) -> GetModulesEvent:
+        """Get modules step."""
+        # get input data, store llm into ctx
+        task = ev.get("task")
+        llm: LLM = ev.get("llm")
```
Comment (logan-markewich) [inline llama-index-packs/llama-index-packs-self-discover/llama_index/packs/self_discover/base.py:107]: For workflows with a single entry point, might be more friendly to confirm these are actually properly passed in, and raise an appropriate error if not

**Review by logan-markewich — COMMENTED** [review]
Looks good to me! Mostly just comments about updating `asyncio.run()` -> `asyncio_run()` and bumping the version of the `llama-index-core` dep

File: llama-index-packs/llama-index-packs-self-discover/llama_index/packs/self_discover/base.py [hunk 7]
```diff
@@ -49,96 +58,114 @@
...
+        ctx.data["llm"] = llm
+
+        if task is None or llm is None:
```
Comment (jerryjliu) [inline llama-index-packs/llama-index-packs-self-discover/llama_index/packs/self_discover/base.py:110]: nit: should have this check before assigning llm to ctx.data

File: llama-index-packs/llama-index-packs-corrective-rag/llama_index/packs/corrective_rag/base.py [hunk 8]
```diff
@@ -44,89 +55,161 @@
...
+        """Ingest step (for ingesting docs and initializing index)."""
+        documents: Optional[List[Document]] = ev.get("documents")
+        tavily_ai_apikey: Optional[str] = ev.get("tavily_ai_apikey")
+
+        if any(i is None for i in [documents, tavily_ai_apikey]):
+            return None
+
         llm = OpenAI(model="gpt-4")
-        self.relevancy_pipeline = QueryPipeline(
+        ctx.data["relevancy_pipeline"] = QueryPipeline(
```
Comment (jerryjliu) [inline llama-index-packs/llama-index-packs-corrective-rag/llama_index/packs/corrective_rag/base.py:94]: this...is probably ok for now until we have hierarchical workflows? @logan-markewich thoughts?

File: llama-index-packs/llama-index-packs-corrective-rag/llama_index/packs/corrective_rag/base.py (hunk 2 above)
Comment (jerryjliu) [reply llama-index-packs/llama-index-packs-corrective-rag/llama_index/packs/corrective_rag/base.py:108]: let's add a check?

**Review by jerryjliu — APPROVED** [review]

File: llama-index-packs/llama-index-packs-corrective-rag/llama_index/packs/corrective_rag/base.py (hunk 8 above)
Comment (logan-markewich) [reply llama-index-packs/llama-index-packs-corrective-rag/llama_index/packs/corrective_rag/base.py:94]: I mean, we could reimplement the pipeline to be a workflow now too

Comment (logan-markewich) [reply llama-index-packs/llama-index-packs-corrective-rag/llama_index/packs/corrective_rag/base.py:94]: and update with the slightly nicer api when it lands

Comment (jerryjliu) [reply llama-index-packs/llama-index-packs-corrective-rag/llama_index/packs/corrective_rag/base.py:94]: @logan-markewich unrelated but one of the original motivations for QPs were the syntatic sugar for common chains e.g. (prompt -> LLM -> output parsing) 

should we do that for workflows?

**Pull request merged**